File: 112-path-sum/path-sum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {boolean}
 */
var hasPathSum = function(root, targetSum) {
    if (root === null) return false
    
    if (root.left === null && root.right === null) {
        return targetSum === root.val
    }

    const remainingSum = targetSum - root.val

    return (
        (root.left !== null && hasPathSum(root.left, remainingSum)) ||
        (root.right !== null && hasPathSum(root.right, remainingSum))
    )
};

File: 112-path-sum/path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        def check_path(node: TreeNode, remaining: int) -> bool:
            if not node:
                return False

            current_remaining = remaining - node.val

            if not node.left and not node.right:
                return current_remaining == 0

            return (check_path(node.left, current_remaining) or
                    check_path(node.right, current_remaining))

        return check_path(root, targetSum)